pages/chat.html

```diff
@@ -0,0 +1,194 @@
+      const [ws, setWs] = React.useState(null);
+
+      React.useEffect(() => {
+        const ws = new WebSocket(prop.url);
+        setWs(ws);
+
+        ws.onopen = () => {
+          console.log('WebSocket connection opened');
+          prop.onOpen()
+          prop.onChannelOpened(function (message) {
+            if (ws.readyState === 1) {
+              ws.send(message);
+            }
+          })
+        };
+
+        ws.onmessage = (event) => {
+          console.log('WebSocket message received:', event.data);
+          try {
+            const json = JSON.parse(event.data);
+            switch (json.type) {
+              case 'accepted':
+              case 'message':
+                prop.onMessage(json)
+                break;
+            }
+          } catch (e) {
+            console.error('Invalid JSON: ', event.data);
+          }
+        };
+
+        ws.onclose = function (event) {
+          console.log('WebSocket connection closed', ws.readyState, event);
+          prop.onClose()
+        };
+
+        ws.onerror = function (error) {
+          console.log('WebSocket error: ', error);
+          if (ws.readyState === 1) {  // open
+            prop.onError(error)
+            ws.close(error);
+          }
+        };
+
+        return function cleanup() {
+          console.log('WebSocket connection cleanup');
+          // ws.close();
+        };
+      }, []);
+
+      return false
+    }
+
+    function App() {
+      const [connecting, setConnecting] = React.useState(true);
+      const [status, setStatus] = React.useState('');
+      const [messages, setMessages] = React.useState([]);
+      const [channel, setChannel] = React.useState(null)
+
+      function submitMessage(event) {
+        event.preventDefault();
+        const data = new FormData(event.currentTarget);
+        const message = data.get('chat-message');
+        if (message && connecting === false && channel) {
+          channel(JSON.stringify({
+            type: 'message',
+            data: message,
+          }))
+        }
+      };
+
+      return (
+        <Container component="main" maxWidth="xs">
+          <CssBaseline />
+          <Box
+            sx={{
+              marginTop: 8,
+              display: 'flex',
+              flexDirection: 'column',
+              alignItems: 'center',
+            }}
+          >
+            <Typography component="h1" variant="h5">
+              Fastify WebSockets React Chat
+            </Typography>
+            <WsConnection
+              url="ws://localhost:3000/chat"
+              onOpen={() => setConnecting(false)}
+              onChannelOpened={(inChannel) => { setChannel(() => { return inChannel }) }}
+              onMessage={({data}) => setMessages((prevState) => { return [...prevState, data] })}
+              onClose={() => { setConnecting(true); setStatus('Disconnected') }}
+              onError={(error) => { setConnecting(true); setStatus(error.message) }}
```

Comment:
One of the reasons the `useEffect` inside `WsConnection` was unexpectedly triggered could have been related to these functions. On every render, 5 new lambda functions are created triggering the `useEffect` if they are inside the dependency array. The common but verbose solution is to use a `useCallback` for each function